### PR TITLE
Support user provided "scope" in JWT and GDC

### PR DESCRIPTION
**Pull request by yihuazhang**
The current C-core JWT token API (i.e., `jwt_credenitals.h` and `jwt_credentials.cc` ) does not support the "scope" field in the token creation API and this PR fixes it. Note that when the JWT token is used together with GDC, we clear the audience field in the token, which is dictated by https://google.aip.dev/auth/4111#scope-vs-audience. 

F.Y.I. I am also working on the gRFC that explains the surface API change in this PR.

---
This change is [Reviewable](https://reviewable.io/reviews/grpc/grpc/26577)

### CODE REVIEW

File: include/grpc/grpc_security.h [hunk 1]
```diff
@@ -324,11 +332,16 @@ GRPCAPI gpr_timespec grpc_max_auth_token_lifetime(void);
...
+   - should_clear_audience dictating the clearance of audience field when it
+     is set to a non-zero value AND user_provided_scope is not nullptr.
+ */
 GRPCAPI grpc_call_credentials*
-grpc_service_account_jwt_access_credentials_create(const char* json_key,
-                                                   gpr_timespec token_lifetime,
-                                                   void* reserved);
+grpc_service_account_jwt_access_credentials_create(
+    const char* json_key, gpr_timespec token_lifetime,
+    const char* user_provided_scope, int should_clear_audience);
```
Comment (shinfan) [inline include/grpc/grpc_security.h:344]: We should rename `should_clear_audience` to `enable_scope_for_jwt` which is defined in the [aip](google.aip.dev/auth/4111).

File: src/core/lib/security/credentials/jwt/json_token.cc [hunk 2]
```diff
@@ -175,15 +176,19 @@ static char* encoded_jwt_claim(const grpc_auth_json_key* json_key,
 
   Json::Object object = {
       {"iss", json_key->client_email},
-      {"aud", audience},
       {"iat", now.tv_sec},
       {"exp", expiration.tv_sec},
   };
+
   if (scope != nullptr) {
```
Comment (shinfan) [inline src/core/lib/security/credentials/jwt/json_token.cc:183]: How about having something like this:
```
if (scope != nullptr && enable_scope_for_jwt) {
  object["scope"] = scope
} else {
  object["aud"] = audience;
}
```

File: include/grpc/grpc_security.h (hunk 1 above)
Comment (yihuazhang) [reply include/grpc/grpc_security.h:344]: Note that this API is not meant to be used for only accessing the google API. Only the jwt token created inside `google_default_credentials_create()` will be used to access the google API. So `enable_scope_for_jwt` does not capture what `should_clear_audience` currently implies (i.e., clearing the audience field if it is set to true).

Comment (yihuazhang) [reply include/grpc/grpc_security.h:344]: So in this PR, we plan to support the following use cases: 
-  allow to specify both `audience` and `scope` fields in a JWT token, and it is controlled by `should_clear_audience = false`. 
- allow to include only `scope` field in a JWT token dictated by https://google.aip.dev/auth/4111, which will solve the regional endpoint issue, and it is controlled by `should_clear_audience = true`. 

It is not clear which case it refers to if using `enable_scope_for_jwt`. WDYT?

File: src/core/lib/security/credentials/jwt/json_token.cc (hunk 2 above)
Comment (yihuazhang) [reply src/core/lib/security/credentials/jwt/json_token.cc:183]: By doing that, it does not allow to have both `scope` and `audience` fields. Please see my above comment.

**Review by markdroth — CHANGES_REQUESTED** [review]
This looks reasonable overall.  My comments are mostly minor, but I would like to better understand both of the use-cases here.

Please let me know if you have any questions.  Thanks!

> Reviewed 17 of 17 files at r1.
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/26577#-:-MdP56303tFewfraSPc7:bi32qtf)* status: all files reviewed, 14 unresolved discussions (waiting on @apolcyn, @shinfan, and @yihuazhang)

___
*[include/grpc/grpc_security.h, line 344 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP1D6AunBJIz7eFvvT-r1-344:-MdP2Ew4CzSNjnOkXndN:b-up7krm) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/include/grpc/grpc_security.h#L344)):*
<details><summary><i>Previously, yihuazhang (yihuaz) wrote…</i></summary><blockquote>

So in this PR, we plan to support the following use cases: 
-  allow to specify both `audience` and `scope` fields in a JWT token, and it is controlled by `should_clear_audience = false`. 
- allow to include only `scope` field in a JWT token dictated by https://google.aip.dev/auth/4111, which will solve the regional endpoint issue, and it is controlled by `should_clear_audience = true`. 

It is not clear which case it refers to if using `enable_scope_for_jwt`. WDYT? 
  
</blockquote></details>

What is the use-case for setting a user-provided scope and *not* clearing the audience field?

Also, if we keep the current semantics, please drop the `should_` prefix and just call it `clear_audience`.  (Same thing throughout the code.)
___
*[include/grpcpp/security/credentials.h, line 215 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP2hsn1EwJ08Nwsm6t:-MdP2hsn1EwJ08Nwsm6u:b-nxe2im) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/include/grpcpp/security/credentials.h#L215)):*
> ```objc
> std::shared_ptr<CallCredentials> ServiceAccountJWTAccessCredentials(
>     const grpc::string& json_key,
>     long token_lifetime_seconds = kMaxAuthTokenLifetimeSecs,
> ```

Would it make sense to have a new variant of `ServiceAccountJWTAccessCredentials()` that takes an option struct?  That way, callers can set just the options they care about and not worry about the order of the parameters.
___
*[src/core/lib/security/credentials/google_default/google_default_credentials.cc, line 257 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP269q7QZaKXwBdluj:-MdP269q7QZaKXwBdluk:b6aoc56) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/src/core/lib/security/credentials/google_default/google_default_credentials.cc#L257)):*
> ```cpp
>         grpc_service_account_jwt_access_credentials_create_from_auth_json_key(
>             key, grpc_max_auth_token_lifetime(), user_provided_scope,
>             1 /* should_clear_audience */);
> ```

Please use `true` here.
___
*[src/core/lib/security/credentials/jwt/json_token.h, line 67 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP3RYZ8Wll2YB7Q2US:-MdP3RYZ8Wll2YB7Q2UT:b-vky2xh) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/src/core/lib/security/credentials/jwt/json_token.h#L67)):*
> ```objc
>                                const char* audience,
>                                gpr_timespec token_lifetime, const char* scope,
>                                int should_clear_audience);
> ```

This should be `bool`.
___
*[src/core/lib/security/credentials/jwt/json_token.cc, line 169 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP3Tsj6_liJmPA8oYk:-MdP3Tsj6_liJmPA8oYl:b-tr282t) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/src/core/lib/security/credentials/jwt/json_token.cc#L169)):*
> ```cpp
>                                const char* audience,
>                                gpr_timespec token_lifetime, const char* scope,
>                                int should_clear_audience) {
> ```

`bool`
___
*[src/core/lib/security/credentials/jwt/json_token.cc, line 273 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP3fK---mswWOAcrdN:-MdP3d4NAxc1zkyrqx5u:b-tr282t) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/src/core/lib/security/credentials/jwt/json_token.cc#L273)):*
> ```cpp
>                                const char* audience,
>                                gpr_timespec token_lifetime, const char* scope,
>                                int should_clear_audience) {
> ```

`bool`
___
*[src/core/lib/security/credentials/jwt/jwt_credentials.h, line 39 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP3jRe3RrpKUBARaof:-MdP3jRe3RrpKUBARaog:b-tr282t) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/src/core/lib/security/credentials/jwt/jwt_credentials.h#L39)):*
> ```objc
>                                               gpr_timespec token_lifetime,
>                                               std::string user_provided_scope,
>                                               int should_clear_audience);
> ```

`bool`
___
*[src/core/lib/security/credentials/jwt/jwt_credentials.h, line 78 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP3lDY9-hDXMyDtzW6:-MdP3lDY9-hDXMyDtzW7:b-tr282t) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/src/core/lib/security/credentials/jwt/jwt_credentials.h#L78)):*
> ```objc
>   gpr_timespec jwt_lifetime_;
>   std::string user_provided_scope_;
>   int should_clear_audience_;
> ```

`bool`
___
*[src/core/lib/security/credentials/jwt/jwt_credentials.h, line 86 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP3n4s-4AKLE4FmVWz:-MdP3n4s-4AKLE4FmVX-:b-tr282t) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/src/core/lib/security/credentials/jwt/jwt_credentials.h#L86)):*
> ```objc
> grpc_service_account_jwt_access_credentials_create_from_auth_json_key(
>     grpc_auth_json_key key, gpr_timespec token_lifetime,
>     std::string user_provided_scope, int should_clear_audience);
> ```

`bool`
___
*[src/core/lib/security/credentials/jwt/jwt_credentials.cc, line 125 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP41OF9vzVPlAe3xwg:-MdP41OF9vzVPlAe3xwh:b-tr282t) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/src/core/lib/security/credentials/jwt/jwt_credentials.cc#L125)):*
> ```cpp
>                                                 gpr_timespec token_lifetime,
>                                                 std::string user_provided_scope,
>                                                 int should_clear_audience)
> ```

`bool`
___
*[src/core/lib/security/credentials/jwt/jwt_credentials.cc, line 145 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP45eI3O7MGNUS0TFq:-MdP45eI3O7MGNUS0TFr:b-tr282t) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/src/core/lib/security/credentials/jwt/jwt_credentials.cc#L145)):*
> ```cpp
> grpc_service_account_jwt_access_credentials_create_from_auth_json_key(
>     grpc_auth_json_key key, gpr_timespec token_lifetime,
>     std::string user_provided_scope, int should_clear_audience) {
> ```

`bool`
___
*[src/core/lib/security/credentials/jwt/jwt_credentials.cc, line 187 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP4AZV8u1CEXLRhRQu:-MdP4AZV8u1CEXLRhRQv:b-gdv234) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/src/core/lib/security/credentials/jwt/jwt_credentials.cc#L187)):*
> ```cpp
>   grpc_core::ExecCtx exec_ctx;
>   if (user_provided_scope == nullptr) user_provided_scope = "";
> ```

Please remove blank lines within functions.
___
*[test/core/security/credentials_test.cc, line 3405 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP4szIEvl6Yb5EiRr2:-MdP4szIEvl6Yb5EiRr3:bfliha8) ([raw file](https://github.com/grpc/grpc/blob/0b65c34ae511d8f4c4f14342bf17a396e109e8ff/test/core/security/credentials_test.cc#L3405)):*
> ```cpp
>   test_get_well_known_google_credentials_file_path();
>   test_channel_creds_duplicate_without_call_creds();
>   test_auth_metadata_context();
> ```

Why are these three lines being removed?  This will effectively disable some tests.

**Review by yihuazhang — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/26577)* status: 16 of 17 files reviewed, 14 unresolved discussions (waiting on @apolcyn, @markdroth, and @shinfan)

___
*[include/grpc/grpc_security.h, line 344 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP1D6AunBJIz7eFvvT-r1-344:-MdP6NTO7gLnf_SsT3s4:bvc9z9d) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/include/grpc/grpc_security.h#L344)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

What is the use-case for setting a user-provided scope and *not* clearing the audience field?

Also, if we keep the current semantics, please drop the `should_` prefix and just call it `clear_audience`.  (Same thing throughout the code.)
</blockquote></details>

In general, a JWT token can contain both audience and scope fields because they represent different meanings. `scope` is the concept introduced by Oauth 2.0, which defines the specific actions applications are allowed to do on a user's behalf. `audience` represents the recipient of the token that consumes it. 

The reason for introducing `clear_audience` is solely due to the requirement specified in https://google.aip.dev/auth/4111 which demands `audience` and `scope` fields cannot co-exist.
___
*[include/grpcpp/security/credentials.h, line 215 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP2hsn1EwJ08Nwsm6t:-MdP8HPuCg53nzrZoXsQ:bw872pn) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/include/grpcpp/security/credentials.h#L215)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Would it make sense to have a new variant of `ServiceAccountJWTAccessCredentials()` that takes an option struct?  That way, callers can set just the options they care about and not worry about the order of the parameters.
</blockquote></details>

Note that we have already included all the fields that need to be specified (i.e., `iss`, `iat`, `exp`) when creating a JWT token in the parameter list (since we provide scope specification, there will be no need to support audience overriding). Also unlike advanced TLS, JWT token is pretty straightforward in a sense that it does not need the advanced features (e.g., credential reloading). So I do not see a future need that will further increase the size of parameter list. 
___
*[src/core/lib/security/credentials/google_default/google_default_credentials.cc, line 257 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP269q7QZaKXwBdluj:-MdPAMucBW-benYSTbHe:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/src/core/lib/security/credentials/google_default/google_default_credentials.cc#L257)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Please use `true` here.
</blockquote></details>

Done.
___
*[src/core/lib/security/credentials/jwt/json_token.h, line 67 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP3RYZ8Wll2YB7Q2US:-MdSoWnp62Fs3PqPqxhw:bb6un1j) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/src/core/lib/security/credentials/jwt/json_token.h#L67)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

This should be `bool`.
</blockquote></details>

Done
___
*[src/core/lib/security/credentials/jwt/json_token.cc, line 169 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP3Tsj6_liJmPA8oYk:-MdSoeM--H4zS8SzN1KB:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/src/core/lib/security/credentials/jwt/json_token.cc#L169)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

`bool`
</blockquote></details>

Done.
___
*[src/core/lib/security/credentials/jwt/json_token.cc, line 273 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP3fK---mswWOAcrdN:-MdSolOJ3twcNTt-qL8D:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/src/core/lib/security/credentials/jwt/json_token.cc#L273)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

`bool`
</blockquote></details>

Done.
___
*[src/core/lib/security/credentials/jwt/jwt_credentials.h, line 39 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP3jRe3RrpKUBARaof:-MdSoztPEOafSkyMKKe8:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/src/core/lib/security/credentials/jwt/jwt_credentials.h#L39)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

`bool`
</blockquote></details>

Done.
___
*[src/core/lib/security/credentials/jwt/jwt_credentials.h, line 78 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP3lDY9-hDXMyDtzW6:-MdSp2X13JRSxWJ90GGp:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/src/core/lib/security/credentials/jwt/jwt_credentials.h#L78)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

`bool`
</blockquote></details>

Done.
___
*[src/core/lib/security/credentials/jwt/jwt_credentials.h, line 86 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP3n4s-4AKLE4FmVWz:-MdSp4pK3nHbivUETQDa:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/src/core/lib/security/credentials/jwt/jwt_credentials.h#L86)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

`bool`
</blockquote></details>

Done.
___
*[src/core/lib/security/credentials/jwt/jwt_credentials.cc, line 125 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP41OF9vzVPlAe3xwg:-MdSpFVZ7Ctjwr0jg8iU:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/src/core/lib/security/credentials/jwt/jwt_credentials.cc#L125)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

`bool`
</blockquote></details>

Done.
___
*[src/core/lib/security/credentials/jwt/jwt_credentials.cc, line 145 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP45eI3O7MGNUS0TFq:-MdSpLYaALLDD8LAu8Pp:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/src/core/lib/security/credentials/jwt/jwt_credentials.cc#L145)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

`bool`
</blockquote></details>

Done.
___
*[src/core/lib/security/credentials/jwt/jwt_credentials.cc, line 187 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP4AZV8u1CEXLRhRQu:-MdSpQT21bFrgNHrYwR5:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/5f80cee54bdc5998d54f42ac00795801059b681f/src/core/lib/security/credentials/jwt/jwt_credentials.cc#L187)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Please remove blank lines within functions.
</blockquote></details>

Done.
___
*[test/core/security/credentials_test.cc, line 3405 at r1](https://reviewable.io/reviews/grpc/grpc/26577#-MdP4szIEvl6Yb5EiRr2:-MdSpS8P1-uquHHh5R-V:b1s0fqu) ([raw file](https://github.com/grpc/grpc/blob/0b65c34ae511d8f4c4f14342bf17a396e109e8ff/test/core/security/credentials_test.cc#L3405)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Why are these three lines being removed?  This will effectively disable some tests.
</blockquote></details>

Opps, my bad. The three lines are added now.

**Review by markdroth — APPROVED** [review]
Looks good!

> Reviewed 11 of 11 files at r3.
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/26577#-:-MdT4De84-yj10dos7vX:b-vs1prh)* status: all files reviewed, 2 unresolved discussions (waiting on @apolcyn and @shinfan)

**Comment by yihuazhang** [comment]
Thanks Mark for the review!

File: include/grpc/grpc_security.h (hunk 1 above)
Comment (shinfan) [reply include/grpc/grpc_security.h]: I see, that makes sense.

**Review by shinfan — APPROVED** [review]
Thanks for the quick fix.

**Comment by yihuazhang** [comment]
Known issues: https://github.com/grpc/grpc/issues/24442, https://github.com/grpc/grpc/issues/26072

**Comment by yihuazhang** [comment]
Thanks for the review, Shin!

**Pull request merged**